add token labelling

joshuawe opened the pull request.

joshuawe commented:
Are the files that have been added stored in a good location?
I think not.

menamerai commented:
@joshuawe I think you need to add `spacy` to the `requirements.txt`

jettjaniak commented on src/delphi/eval/token_labelling.py:
```diff
@@ -0,0 +1,184 @@
+"""
+
+from pprint import pprint
+from typing import List, Optional
```
you don't need to import `List` for typing, you can just use `list`

joshuawe replied:
You are right.
But unfortunately my venv was created using Python 3.8 where `List` is required.

> In Python 3.8 and earlier versions, to specify the type of elements within containers like lists, you need to import List from the typing module. This is because the built-in list type was not subscriptable (you couldn't use list[int] directly for type hints).

jettjaniak replied:
please update to 3.10, you'll run into issues when we use features from 3.9+ (I think we already do?)

jettjaniak replied:
LMK if you need help on this

joshuawe replied:
Done

jettjaniak commented on src/delphi/eval/token_labelling.py:
```diff
@@ -0,0 +1,184 @@
+            print("   ", label.ljust(10), key)
+
+
+def label_single_token(token: Token) -> List:
```
I would like to have function that just takes a `str` token label. Should we return this as a dict?

jettjaniak reviewed (commented):
Let's make a pickle with dict[int, dict[str, bool]] that labels every single token in the tokenizer.

joshuawe commented:
Have a look at the notebook, where function calling as well as the creation of the pickle object which is basically Look Up Table containing a dict with keys *token_id* and value *token_labels*.

Note that *token_labels* is itself a dict with all keys *label_name* and values *bool*

For example:
```
token_ids_lablled = {
  0: {'Starts with space': False, 'Capitalized': True, 'Is Noun': True, 'Is Pronoun': False, 'Is Adjective': False, 'Is Verb': False, 'Is Adverb': False, 'Is Preposition': False, 'Is Conjunction': False, 'Is Interjunction': False, 'Is Named Entity': False},
  1: {'Starts with space': False, 'Capitalized': True, 'Is Noun': True, 'Is Pronoun': False, 'Is Adjective': False, 'Is Verb': False, 'Is Adverb': False, 'Is Preposition': False, 'Is Conjunction': False, 'Is Interjunction': False, 'Is Named Entity': False},
  2: ...
...
}
```

The question that remains is 
1. where to store the pickle. 
2. How to make it accessible. Should it always be loaded into the memory, when the module is imported?

joshuawe commented:
Another interesting thing is the following, 
the vocab size of the tokenizer is larger than 4000 by far.
```python
>>> from transformers import AutoTokenizer

>>> tokenizer = AutoTokenizer.from_pretrained("roneneldan/TinyStories-1M")

>>> vocab_size = tokenizer.vocab_size
>>> print("The vocab size is:", vocab_size)
50247
``` 

So, am I using the correct tokenizer? And are there multiple tokenizers?

jettjaniak commented:
The tokenizer is wrong, it's from the original tiny stories models. You need to pass model name from any of the models on delphi-suite Hugging Face to use our small tokenizer that was trained on tiny stories dataset.

jettjaniak commented on requirements.txt:
```diff
@@ -9,4 +9,6 @@ black==23.12.1
\ No newline at end of file
+isort==5.13.2
+spacy==3.7.2
+chardet==5.2.0
```
what is it?

joshuawe replied:
`spacy` -> for LLM tasks such as analysis of tokens (part of speach POS)
`chardet` -> a dependency required for `spacy` that was not installed automatically

jettjaniak commented on requirements.txt:
```diff
@@ -1,4 +1,4 @@
-torch==2.1.2
+# torch==2.1.2
```
?

joshuawe replied:
ah damn it, sorry, mistake from my side.

jettjaniak commented on src/delphi/__init__.py:
```diff
@@ -1,3 +1,5 @@
 from beartype.claw import beartype_this_package  # <-- hype comes
 
+from . import eval
+
```
why?

joshuawe replied:
I still require this so I can import everything under `delphi`. For WHATEVER reason an empty `__init__.py` file will not allow me to import it. (of course, I used `pip install -e .`)

jettjaniak commented on src/delphi/eval/token_labelling.py:
```diff
@@ -0,0 +1,211 @@
+}
+
+
+def explain_Token_labels(token: Optional[Token] = None) -> None:
```
lowercase please

jettjaniak commented on src/delphi/eval/token_labelling.py:
```diff
@@ -0,0 +1,211 @@
+    return labels
+
+
+def label_sentence(tokens: Union[Doc, list[Token]]) -> list[dict[str, bool]]:
```
```suggestion
def label_sentence(tokens: Doc | list[Token]) -> list[dict[str, bool]]:
```
Union is deprecated

jettjaniak commented on src/delphi/eval/token_labelling.py:
```diff
@@ -0,0 +1,211 @@
+        corresponding token length where each entry provides the labels/categories
+        for the token. Sentence -> Token -> Labels
+    """
+    assert isinstance(sentences, list)
```
Redundant, we're running beartype. Remove it and try passing something different than specified type, it should fail.

jettjaniak commented on src/delphi/eval/token_labelling.py:
```diff
@@ -0,0 +1,211 @@
+    tokenized: bool = True,
+    verbose: bool = False,
+) -> list[list]:
+    """
+    Labels tokens in a sentence batchwise. Takes the context of the token into
+    account for dependency labels (e.g. subject, object, ...).
+
+    Parameters
+    ----------
+    sentences : list
+        A batch/list of sentences, each being a list of tokens.
+    tokenized : bool, optional
+        Whether the sentences are already tokenized, by default True. If the sentences
+        are full strings and not lists of tokens, then set to False. If true then `sentences` must be list[list[str]].
+    verbose : bool, optional
+        Whether to print the tokens and their labels to the console, by default False.
+
+    Returns
+    -------
+    list[list[dict[str, bool]]
+        Returns a list of sentences. Each sentence contains a list of its
+        corresponding token length where each entry provides the labels/categories
+        for the token. Sentence -> Token -> Labels
+    """
```
My preferred docstring template is 

"""Short one-line description

Optional longer description
"""

If you want to list and describe all arguments that's fine, but specifying their type and optionality is redundant, you can see all of that in function definition.

jettjaniak commented on src/delphi/eval/token_labelling.py:
```diff
@@ -0,0 +1,211 @@
+
+        # go through each token in the sentence
+        for token, labelled_token in zip(doc, labelled_tokens):
+            # labelled_token = label_single_Token(token)
+            # labels = list()  #  The list holding labels of a single token
+            # for _, category_check in TOKEN_LABELS.items():
+            #     label = category_check(token)
+            #     labels.append(label)
+            # add current token's to the list
+            # labelled_tokens.append(labelled_token)
```
?

jettjaniak commented:
you need to rebase on top of main

jettjaniak commented:
The notebook is good as an explanation, but generating the labels shouldn't require running it. Let's add something in scripts/ for that.

jettjaniak commented:
Is there any reason we don't have a function like `label_token(token: str) -> dict[str, bool]`? Combining random tokens into sentences seems opaque and bug prone.

jettjaniak commented:
for now just store the pickle in eval directory

jettjaniak commented:
We should debug this. What is your exact setup? System, python interpreter,
how you make your venv etc. Also, can you `import sys; print(sys.path)`

On Fri, 9 Feb 2024, 08:39 Joshua We, ***@***.***> wrote:

> ***@***.**** commented on this pull request.
> ------------------------------
>
> In src/delphi/__init__.py
> <https://github.com/delphi-suite/delphi/pull/21#discussion_r1484559657>:
>
> > +from . import eval
> +
>
> I still require this so my I can import everything. For WHATEVER reason an
> empty *init*.py file will not allow me to import it. (of course, I used pip
> install -e .)
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/delphi-suite/delphi/pull/21#discussion_r1484559657>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AECKP3YGYAEMHVD3A4UROPDYSZGLNAVCNFSM6AAAAABCW6SGQOVHI2DSMVQWIX3LMV43YUDVNRWFEZLROVSXG5CSMV3GSZLXHMYTQNZSHAYDAMRVGU>
> .
> You are receiving this because you commented.Message ID:
> ***@***.***>
>

joshuawe commented:
> We should debug this. What is your exact setup? System, python interpreter, how you make your venv etc. Also, can you `import sys; print(sys.path)`

My setup:
- Windows
- conda for venv creation
- `pip install -e .`
- Python interpreter by running `where Python`, the first entry is : 
  `C:/Users/.../anaconda3/envs/delphi2/python.exe`


Here the other outputs
```Python
>>> print(dir(delphi))
['__builtins__', '__cached__', '__doc__', '__file__', '__loader__', '__name__', '__package__', '__path__', '__spec__', 'beartype_this_package', 'eval']

>>> import sys; print(sys.path)
['c:\\"path to \\delphi\\notebooks', 'c:\\Users\\...\\anaconda3\\envs\\delphi2\\python310.zip', 'c:\\Users\\...\\anaconda3\\envs\\delphi2\\DLLs', 'c:\\Users\\...\\anaconda3\\envs\\delphi2\\lib', 'c:\\Users\\...\\anaconda3\\envs\\delphi2', '', 'c:\\Users\\...\\anaconda3\\envs\\delphi2\\lib\\site-packages', 'c:\\users\\...\\joshua\\research\\2024_01_tinyevals\\delphi\\src', 'c:\\Users\\...\\anaconda3\\envs\\delphi2\\lib\\site-packages\\win32', 'c:\\Users\\...\\anaconda3\\envs\\delphi2\\lib\\site-packages\\win32\\lib', 'c:\\Users\\...\\anaconda3\\envs\\delphi2\\lib\\site-packages\\Pythonwin']
```

jettjaniak commented:
The only thing that comes to mind is conda being weird. You could just download and install python for windows and make venv this way

joshuawe commented:
- added tests
- added a script `label_all_tokens.py` that does some token labelling
- now two files are saved under `src/delphi/eval/`. using the tokenizer from a given huggingface language model repo
  1. the tokens and their ids are stored in `all_tokens_{huggingface_model_name}.txt`
  2. The labelled tokens stored as `labelled_token_ids_dict_{huggingface_model_name}.pkl`
- had to update requirements in order to extract/run the delphi tokenizer

Small todo's left for tomorrow:
- [ ] Clear up the `__init__.py` files again
- [x] Implement the function that takes in a token as a `str` and checks for its labels in the token_label_dict
  @jettjaniak Should the tokenized dict be loaded in memory *always*, i.e. when the module is loaded it also loads the pickle?

jettjaniak commented:
> the tokens and their ids are stored in all_tokens_{huggingface_model_name}.txt
The labelled tokens stored as labelled_token_ids_dict_{huggingface_model_name}.pkl

Can we drop the _{huggingface_model_name}? All delphi models use the same tokenizer.

>  Implement the function that takes in a token as a str and checks for its labels in the token_label_dict
@jettjaniak Should the tokenized dict be loaded in memory always, i.e. when the module is loaded it also loads the pickle?

I assume this question only applies if you're implementing this function? I'd say don't implement it, at least for now. The pickle is enough for the demo. And in this case, you should never be loading the pickle in code included in this PR.

jettjaniak commented:
branch is out-of-date with main, you need to `git rebase main`

jettjaniak commented:
were you using `git merge`? something is wrong with "Files changed" tab in this PR and with the commit history

jettjaniak commented on src/delphi/eval/token_labelling.py:
```diff
@@ -0,0 +1,198 @@
+"""
+This script creates labels for tokens in a sentence. 
+It takes the context of the token into account. 
+Additionally, it can visualize the sentences and their poart-of-speech (POS) tags.
+"""
```
This is not a script, but a module - collection of importable functions. They should have their own docstrings, but file docstring is not necessary.

jettjaniak commented on src/delphi/eval/token_labelling.py:
```diff
@@ -0,0 +1,198 @@
+
+from typing import Callable, Optional
+
+import spacy  # pylint: disable=import-error
+from spacy.tokens import Doc  # pylint: disable=import-error
```
can you remove the # pylint comments now?

jettjaniak commented on src/delphi/eval/token_labelling.py:
```diff
@@ -0,0 +1,198 @@
+        Returns a dictionary with the token's labels as keys and their
+        corresponding boolean values.
+    """
+    assert isinstance(token, Token)
```
beartype should take care of this automatically

jettjaniak commented on src/delphi/eval/token_labelling.py:
```diff
@@ -0,0 +1,198 @@
+        corresponding boolean values.
+    """
+    assert isinstance(token, Token)
+    labels = dict()  #  The list holding labels of a single token
```
not a list

jettjaniak commented on tests/eval/test_token_labelling.py:
```diff
@@ -0,0 +1,114 @@
+
+
+@pytest.fixture
+def create_dummy_doc() -> tuple[str, Doc, dict[str, bool]]:
```
fixture should probably be named dummy_doc() instead of create_dummy_doc()

jettjaniak commented on tests/eval/test_token_labelling.py:
nice!

jettjaniak commented on scripts/label_all_tokens.py:
```diff
@@ -0,0 +1,116 @@
+def main():
+    print("\n", " LABEL ALL TOKENS ".center(50, "="), "\n")
+    # Access command-line arguments
+    args = sys.argv[1:]
```
please use argparse

jettjaniak commented on scripts/label_all_tokens.py:
```diff
@@ -0,0 +1,116 @@
+    print("(1) Create a list of all tokens in the tokenizer's vocabulary ...")
+
+    # Load the tokenizer from Huggingface
+    tokenizer = AutoTokenizer.from_pretrained(model_name)
```
is this the line that requires sentencepiece and protobuf?

joshuawe replied:
yes exactly

jettjaniak commented on scripts/label_all_tokens.py:
```diff
@@ -0,0 +1,116 @@
+        # create a batch of token_ids
+        end = min(start + batch_size, max_token_id)
+        token_ids = list(range(start, end))
+        # decode the token_ids to get a list of tokens, a 'sentence'
+        tokens = decode(tokenizer, token_ids)  # list of tokens == sentence
+        # put the sentence into a list, to make it a batch of sentences
+        sentences = [tokens]
```
This seems wrong. Are you turning a list of unrelated tokens into a single string?

jettjaniak replied:
How are they being split afterwards?

joshuawe replied:
True, it might not make sense to process the tokens batch-wise.
Doing it individually is extremely slow. And as we are NOT looking at dependencies of tokens in a sentence (subject, object, etc.) I assumed it would be fine to increase the computational speed this way.
I had also assumed that we might call this function more frequently.
But as we will use the look-up-dict I can calculate it once but slowly instead. (So not batched)

jettjaniak commented:
did you try venv w/o conda?

joshuawe commented:
ok, everything is set up.
It can definitely be used for demo purposes.
Please checkout the notebook.

What I just realised is that there are a few tokens that seem not to be labelled entirely correctly. I will check this in an upcoming task.

jaidhyani commented:
@joshuawe - would it be okay if I adapted some of this code for the demo without pulling in the whole PR?

joshuawe commented:
@jaidhyani  you can do that, if you think that makes sense.
But why not integrate the PR?
The reason for the previously failed check was that it wants to load a model that is too for the github action. I chose a smaller one, so the check passes. So, if that was the previous aprehension to not merge the PR, it should now be resolved.

jettjaniak commented:
rebase required
